transform_inline.ml

```diff
@@ -0,0 +1,312 @@
+    | Call (x, f, exs) ->
+      Call (replace x, in_expression f, List.map in_arg exs)
+    | Stop e ->
+      Stop (in_expression e)
+    | Return e ->
+      Return (in_expression e)
+    | Decl_var (x, exp) ->
+      Decl_var (replace x, in_expression exp)
+    | Decl_array (x, def) ->
+      let def = match def with
+        | Length e -> Length (in_expression e)
+        | List es -> List (List.map in_expression es)
+      in Decl_array (replace x, def)
+    | Assign (x, exp) ->
+      Assign (replace x, in_expression exp)
+    | Array_assign (x, index, exp) ->
+      Array_assign (replace x, in_expression index, in_expression exp)
+    | Drop x ->
+      Drop (replace x)
+    | Print exp ->
+      Print (in_expression exp)
+    | Assert exp ->
+      Assert (in_expression exp)
+    | Branch (exp, l1, l2) ->
+      Branch (in_expression exp, l1, l2)
+    | Osr {label; cond; target; map} ->
+      let cond = List.map in_expression cond in
+      let map = List.map in_osr map in
+      Osr {label; cond; target; map}
+    | Read x ->
+      Read (replace x)
+    | Label _ | Goto _ | Comment _ ->
+      assert (VarSet.is_empty (used_vars instr));
+      instr
+  in
+
+  (* Given `instructions` and the function identifier, extract information from
+     the first callsite where a call is made to that function *)
+  let extract_callsite instrs fun_ref : (pc * variable * argument list)=
+    let rec at pc =
+      match[@warning "-4"] instrs.(pc) with
+      | Call (x, (Simple (Constant (Fun_ref f))), es) ->
+        if f = fun_ref then (pc, x, es) else at (pc + 1)
+      | _ -> at (pc + 1)
+    in
+    at 0
+  in
+
+  (* Replace variables in `callee` instructions so that they don't match the
+     `caller` variables. The formals list of the callee is also updated
+     accordingly *)
+  let replace_vars caller ({formals; instrs} as callee) =
+    let caller_vars = function_vars caller in
+    let callee_vars = function_vars callee in
+    let replacements =
+      List.map (fun old -> (old, Edit.fresh caller_vars old))
+        (VarSet.elements callee_vars)
+    in
+    {formals = List.map (fun var -> fresh caller_vars var) formals;
+     instrs = Array.map (replace_vars_in_instruction replacements) instrs;}
+  in
+
+  (* Redirect all returns in the `callee` to a unique label (inserted later at
+     the end of the callee's body) after assigning the return expression to a
+     designated `res_var` and dropping all the variables in scope. *)
+  let replace_returns res_var ret_lab ({formals; instrs} as callee) =
+    let extract_returns instrs : (pc * expression) list=
+      let rec loop pc acc =
+        if pc = Array.length instrs then acc
+        else match[@warning "-4"] instrs.(pc) with
+          | Return e -> loop (pc + 1) ((pc, e)::acc)
+          | _ -> loop (pc + 1) acc
+      in
+      loop 0 []
+    in
+    let scope = Scope.infer {formals = VarSet.of_list callee.formals;
+                             instrs = callee.instrs;}
```

Comment:
I guess at this point the scope of the callee is guaranteed to be still correct in itself, because you did just renamings, right? But we got to be extra careful since it will fail on an ill-scoped function.